Export CMake targets

SeanBryan51 opened the pull request.

SeanBryan51 committed:
Define imported targets for find modules

SeanBryan51 committed:
FindNetCDF.cmake: check C language support is enabled if running tests

This change is optional.

SeanBryan51 committed:
LibFind.cmake: add CMAKE_PREFIX_PATH to list of search paths

This change is optional.

SeanBryan51 commented:
TODO
- [x] Test exported targets with PnetCDF build
- [ ] ~~Test exported targets with MPI Serial~~
- [x] Determine minimum CMake version required (last step)

TODO (optional)
- [ ] ~~LibFind.cmake: use `NAME_MISMATCHED` in `find_package_handle_standard_args` to fix the warning "`<pkg>_<comp>` does not match the name of the calling package `<pkg>`."~~
- [ ] ~~Declare dependency on MPI at CMake level. It looks like PIO relies on the MPI compiler wrapper for linking against MPI and does not use `target_link_libraries` to link against MPI. The problem with doing this is that CMake won't know that MPI is required for the `pioc` and `piof` targets. This information should be in the exported targets so that an error can be raised if MPI cannot be found.~~

SeanBryan51 committed:
FindNetCDF.cmake: set NetCDF_FOUND if all components have been found

This is required so that the `find_dependency` macro does not raise a
fatal error when requesting components.

SeanBryan51 commented on src/gptl/CMakeLists.txt:
```diff
@@ -85,7 +98,7 @@ if (ENABLE_LIBRT)
     target_compile_definitions (gptl
       PUBLIC HAVE_LIBRT)
     target_include_directories (gptl
-      PUBLIC ${LIBRT_INCLUDE_DIRECTORIES})
+      PUBLIC "$<BUILD_INTERFACE:${LIBRT_INCLUDE_DIRECTORIES}>")
     target_link_libraries (gptl
       PUBLIC ${LIBRT_LIBRARIES})
```
TODO: this needs to be using an imported target.

anton-seaice commented on cmake/FindGPTL.cmake:
Is it common to change these Find...cmake files ? I thought they were copied from upstream often ?

SeanBryan51 replied:
You're right that it is common for libraries to distribute find modules but I think these find modules for third party libraries were created and are maintained by the PIO developers.

I would consider these to be pretty out of date since modern find modules should already support the use targets rather than variables like `PKG_COMP_LIBRARIES` or `PKG_COMP_INCLUDE_DIRS`. The latter results in a non-relocatable package due to hard coded absolute paths to dependencies in the exported CMake targets.

anton-seaice replied:
Ok cool :)

SeanBryan51 committed:
FindPnetCDF.cmake: set PnetCDF_FOUND if all components have been found

This is required so that the `find_dependency` macro does not raise a
fatal error when requesting components.

SeanBryan51 committed:
PIOConfig.cmake.in: add PNetCDF support

SeanBryan51 committed:
Export CMake targets

These are the smallest required changes to export targets for PIO_C,
PIO_Fortran and PIO_GPTL targets.

Note that with this change, the installed PIO package is not
relocatable. That is, target properties which may contain paths, such as
INTERFACE_INCLUDE_DIRECTORIES and INTERFACE_LINK_LIBRARIES of the
exported targets may contain the absolute paths to libraries and include
directories of external dependencies as found on the machine the package
was made on. If any of these absolute paths are no longer valid on the
machine the package is installed on, then the installation will break.
See [Creating Relocatable Packages][creating-relocatable-packages] for
more details.

[creating-relocatable-packages]: https://cmake.org/cmake/help/latest/manual/cmake-packages.7.html#creating-relocatable-packages

SeanBryan51 committed:
Package find modules with install
